core/src/main/scala/org/http4s/HttpVersion.scala

```diff
@@ -35,20 +35,21 @@ object HttpVersion {
       case "HTTP/1.1" => Either.right(`HTTP/1.1`)
       case "HTTP/1.0" => Either.right(`HTTP/1.0`)
```

Comment:
Historically these two dominate and microbenchmarks showed it was safer to jump straight to them.  Is this still true in cats-parse?

Reply:
It would be interesting to see. I'd be a bit surprised this would be better than an ideal since you are checking a whole string, but if it matches 99% of the time on the first string then probably it is a win.

Reply:
I couldn't beat this even making them specific cases in parboiled2, or using parboiled2's map facility.  No matter how little work the parser did, it couldn't compete with two string comparisons.

Reply:
It would be interesting to see.... The string parser is calling regionMatches with the offset, so there is no internal copying or anything like that.

Reply:
I realized we could do better and cache the `Right` wrapper. [Gist](https://gist.github.com/rossabaker/f97a7f340296575dc6fd90197d7b1cb5)

Results:

```
[info] Benchmark                        Mode  Cnt          Score          Error  Units
[info] ParserBench.parseFlyweight_1_0  thrpt    4   13739316.905 ± 1074228.764  ops/s
[info] ParserBench.parseFlyweight_1_1  thrpt    4   13644878.755 ±  351068.389  ops/s
[info] ParserBench.parseFlyweight_2_0  thrpt    4   13290807.918 ± 1223035.879  ops/s
[info] ParserBench.parseMatch_1_0      thrpt    4  353128170.066 ± 41155378.386  ops/s
[info] ParserBench.parseMatch_1_1      thrpt    4  346638271.637 ± 26582497.893  ops/s
[info] ParserBench.parseMatch_2_0      thrpt    4   13613696.324 ±   251148.799  ops/s
[info] ParserBench.parseOrElses_1_0    thrpt    4   26153858.757 ± 1530188.268  ops/s
[info] ParserBench.parseOrElses_1_1    thrpt    4   31797487.174 ± 7072299.041  ops/s
[info] ParserBench.parseOrElses_2_0    thrpt    4    8470594.820 ±  249313.539  ops/s
[info] ParserBench.parseSimple_1_0     thrpt    4   13589006.294 ±   880396.476  ops/s
[info] ParserBench.parseSimple_1_1     thrpt    4   13429852.948 ±  2284349.941  ops/s
[info] ParserBench.parseSimple_2_0     thrpt    4   13239167.061 ±  1578824.941  ops/s
```

It's possible I'm cheating by parsing the same string over and over again instead of a mix, but I think caching the Right makes sense.  

This is a rare opportunity for optimization.  We won't be doing similar in most of our parsers.

Reply:
Pretty conclusive. Big gap. One note: I would use string1("HTTP/") and not compose the / on with char.

Reply:
Also note you have to allocate the internal mutable State and the result for even the shortest parse. So I guess for this example, checking the likely case is worth it.